elf-util: discard PT_LOAD segment early based on the start address.

Romain-Geissler-1A opened the pull request:
Indeed when iterating over all the PT_LOAD segment of the core dump while trying to look for the elf headers of a given module, we iterate over them all and try to use the first one for which we can parse a package metadata, but the start address is never taken into account, so absolutely nothing guarantees we actually parse the right ELF header of the right module we are currently iterating on.

This specific change was proposed as part of a wider issue we are facing in some corner case scenarios with systemd-coredumper. The whole discussion was initiated here https://sourceware.org/pipermail/elfutils-devel/2023q2/006225.html and this specific change seemed to make sense according to Mark Wielaard (maintainer of the elfutils project): https://sourceware.org/pipermail/elfutils-devel/2023q2/006250.html

CC @bluca who was involved in this elfutils mail thread.

PS: this is my first attempt to contribute something into systemd, so I might have done some wrong things ;)

bluca commented:
Have you tested this by crashing a binary that has the .note.package metadata info?

poettering commented on src/shared/elf-util.c:
```diff
@@ -538,6 +538,10 @@ static int module_callback(Dwfl_Module *mod, void **userdata, const char *name,
                         continue;
 
+                /* This PT_LOAD segment doesn't contain the start address, so it can't be the module we are looking for. */
+                if (start < program_header->p_vaddr || start >= program_header->p_vaddr + program_header->p_memsz)
```
do we need to be afraid of overflows on that addition?

bluca replied:
@Romain-Geissler-1A I missed this open question - can you double check?

bluca replied:
ahah that's funny, now it shows your answer before my question, but the timestamps are inverted :-)

> I guess it all depends on how well we trust the core dump file.

In general we consider core files we analyze to be untrusted, so it's better safe than sorry.

Romain-Geissler-1A replied:
Ah so I have to prevent this overflow indeed.

See #28128.

Romain-Geissler-1A commented:
> Have you tested this by crashing a binary that has the .note.package metadata info?

Honestly, no I didn't test this. Right now I don't have any VMs I would be root on and where I would be able to install a patched systemd, my day to day job involves containers a lot, running on remote servers where I have no root account, but not playing with disposable VMs. I need to discuss internally how to get this in my company, worse case I will try to install some VirtualBox directly on my laptop.

bluca commented:
> > Have you tested this by crashing a binary that has the .note.package metadata info?
> 
> Honestly, no I didn't test this. Right now I don't have any VMs I would be root on and where I would be able to install a patched systemd, my day to day job involves containers a lot, running on remote servers where I have no root account, but not playing with disposable VMs. I need to discuss internally how to get this in my company, worse case I will try to install some VirtualBox directly on my laptop.

Parsing happens on the second stage, so you could simply place a drop-in for systemd-coredump@.service in /run/systemd/system that replaces the ExecStart with one pointing to your local build - no need to install anything

Romain-Geissler-1A commented:
Ok thanks for the tip ;)  I will try to setup this then, and will tell you the result when I will have it.

Romain-Geissler-1A commented:
Ok so I was able to confirm my patch work, and actually it fixes some mistakes of the past implementations. It's mainly for me (as a reminder for later), but I will copy/paste here how exactly I tested and how what I did can be reproduced.

First step is to create a dummy core dump using fedora 37 (different from the distro we will use to analyze the core, to simulate a core happening from inside a container on OpenShift). I crashed `sleep` rather than `bash` on purpose to show later some wrong reports made by the old implementation. Since libc.so doesn't have any package metadata, let's `LD_PRELOAD=libreadline.so.8` to have a module (different from the exec'ed binary) that has a valid package metadata:

```
rgeissler@nceoberhel0058:~/workspace/systemd (core-parsing-use-start-address *%|u=)> podman run -t -i --rm -v $(pwd):$(pwd) -w $(pwd) fedora:37 bash -x -c 'LD_PRELOAD=libreadline.so.8 sleep 1000 & SLEEP_PID="$!" && sleep 1 && kill -11 "${SLEEP_PID}" && mv "core.${SLEEP_PID}" the-core'
```

Second step is to start another container, based on another OS, where we will build/test systemd. I used fedora:38 this time, created with this dockerfile:

```
FROM fedora

RUN dnf install -y vim dnf-plugins-core systemd systemd-udev procps less gdb elfutils 

RUN dnf builddep -y systemd

# Start systemd.
ENTRYPOINT /sbin/init
```

and started like this:

```
podman run -d --rm --privileged -v $(pwd):$(pwd) -w $(pwd) --cidfile amadeus/container.cid $(podman build -q amadeus/)
```

Then from inside this container (ie thanks to: `podman exec -t -i "$(cat amadeus/container.cid)" bash`):
 - Configure (only the first time): `meson setup build`
 - Build: `ninja -C build`
 - Try to run `systemd-coredump` manually, using a fake top level executable (we use `$$` for the PID of the crashing process, which in my case will be a `bash` process, while in real condition it would have been a fedora 37's `sleep` executable, so part of the systemd-coredump logic is expectedly wrong for the executable):

```
rm -rf /var/lib/systemd/coredump/* && journalctl --flush --rotate && journalctl --vacuum-time=1s && cat the-core | build/systemd-coredump $$ 0 0 11 1687360485 1000000000000000 localhost && journalctl
```

I have added some traces of my own to help me debugging.

Without this patch, I get:

```
[root@2d4eefd35cce systemd]# rm -rf /var/lib/systemd/coredump/* && journalctl --flush --rotate && journalctl --vacuum-time=1s && cat the-core | build/systemd-coredump $$ 0 0 11 1687360485 1000000000000000 localhost && journalct
l
Deleted archived journal /var/log/journal/023dff7f5ff0462aae79588e46763ce8/system@a59fd29d8d8c4239aad53aadb8b04c83-0000000000000b08-0005fea70d133ba2.journal (3.5M).
Vacuuming done, freed 3.5M of archived journals from /var/log/journal/023dff7f5ff0462aae79588e46763ce8.
Vacuuming done, freed 0B of archived journals from /run/log/journal.
Vacuuming done, freed 0B of archived journals from /var/log/journal.
Jun 21 17:26:55 2d4eefd35cce systemd-journald[24]: System Journal (/var/log/journal/023dff7f5ff0462aae79588e46763ce8) is 16.0M, max 2.3G, 2.3G free.
Jun 21 17:26:55 2d4eefd35cce systemd-coredump[7402]: Due to PID 1 having crashed coredump collection will now be turned off.
Jun 21 17:26:55 2d4eefd35cce (sd-parse-elf)[7403]: sym_dwfl_module_getelf for /usr/bin/sleep worked
Jun 21 17:26:55 2d4eefd35cce (sd-parse-elf)[7403]: sym_dwfl_module_getelf for /usr/lib64/libtinfo.so.6.3 returned NULL
Jun 21 17:26:55 2d4eefd35cce (sd-parse-elf)[7403]: parse_package_metadata for /usr/lib64/libtinfo.so.6.3, with fallback method which consist in fetching using only the core dump file data returned: 1
Jun 21 17:26:55 2d4eefd35cce (sd-parse-elf)[7403]: sym_dwfl_module_getelf for /usr/lib64/libc.so.6 returned NULL
Jun 21 17:26:55 2d4eefd35cce (sd-parse-elf)[7403]: parse_package_metadata for /usr/lib64/libc.so.6, with fallback method which consist in fetching using only the core dump file data returned: 1
Jun 21 17:26:55 2d4eefd35cce (sd-parse-elf)[7403]: sym_dwfl_module_getelf for /usr/lib64/libreadline.so.8.2 returned NULL
Jun 21 17:26:55 2d4eefd35cce (sd-parse-elf)[7403]: parse_package_metadata for /usr/lib64/libreadline.so.8.2, with fallback method which consist in fetching using only the core dump file data returned: 1
Jun 21 17:26:55 2d4eefd35cce (sd-parse-elf)[7403]: sym_dwfl_module_getelf for /usr/lib64/ld-linux-x86-64.so.2 returned NULL
Jun 21 17:26:55 2d4eefd35cce (sd-parse-elf)[7403]: parse_package_metadata for /usr/lib64/ld-linux-x86-64.so.2, with fallback method which consist in fetching using only the core dump file data returned: 1
Jun 21 17:26:55 2d4eefd35cce (sd-parse-elf)[7403]: sym_dwfl_module_getelf for linux-vdso.so.1 worked
Jun 21 17:26:55 2d4eefd35cce (sd-parse-elf)[7403]: parse_package_metadata for linux-vdso.so.1, with elf from sym_dwfl_module_getelf didn't return anything.
Jun 21 17:26:55 2d4eefd35cce (sd-parse-elf)[7403]: parse_package_metadata for linux-vdso.so.1, with fallback method which consist in fetching using only the core dump file data returned: 0
Jun 21 17:26:55 2d4eefd35cce systemd-coredump[7402]: [🡕] Process 59 (bash) of user 0 dumped core.

                                                     Module /usr/bin/sleep from rpm bash-5.2.15-3.fc38.x86_64
                                                     Module /usr/lib64/libtinfo.so.6.3 from rpm coreutils-9.1-8.fc37.x86_64
                                                     Module /usr/lib64/libc.so.6 from rpm coreutils-9.1-8.fc37.x86_64
                                                     Module /usr/lib64/libreadline.so.8.2 from rpm coreutils-9.1-8.fc37.x86_64
                                                     Module /usr/lib64/ld-linux-x86-64.so.2 from rpm coreutils-9.1-8.fc37.x86_64
                                                     Stack trace of thread 2:
                                                     #0  0x00007ff242961fc7 n/a (/usr/lib64/libc.so.6 + 0xd1fc7)
                                                     #1  0x000000003b975847 n/a (n/a + 0x0)
                                                     ELF object binary architecture: AMD x86-64
```

With this patch, I get:

```
[root@2d4eefd35cce systemd]# rm -rf /var/lib/systemd/coredump/* && journalctl --flush --rotate && journalctl --vacuum-time=1s && cat the-core | build/systemd-coredump $$ 0 0 11 1687360485 1000000000000000 localhost && journalct
l
Vacuuming done, freed 0B of archived journals from /var/log/journal.
Vacuuming done, freed 0B of archived journals from /run/log/journal.
Deleted archived journal /var/log/journal/023dff7f5ff0462aae79588e46763ce8/system@a59fd29d8d8c4239aad53aadb8b04c83-0000000000000b16-0005fea712ffdab0.journal (3.5M).
Vacuuming done, freed 3.5M of archived journals from /var/log/journal/023dff7f5ff0462aae79588e46763ce8.
Jun 21 17:28:14 2d4eefd35cce systemd-journald[24]: System Journal (/var/log/journal/023dff7f5ff0462aae79588e46763ce8) is 16.0M, max 2.3G, 2.3G free.
Jun 21 17:28:14 2d4eefd35cce systemd-coredump[7456]: Due to PID 1 having crashed coredump collection will now be turned off.
Jun 21 17:28:14 2d4eefd35cce (sd-parse-elf)[7457]: sym_dwfl_module_getelf for /usr/bin/sleep worked
Jun 21 17:28:14 2d4eefd35cce (sd-parse-elf)[7457]: sym_dwfl_module_getelf for /usr/lib64/libtinfo.so.6.3 returned NULL
Jun 21 17:28:14 2d4eefd35cce (sd-parse-elf)[7457]: parse_package_metadata for /usr/lib64/libtinfo.so.6.3, with fallback method which consist in fetching using only the core dump file data returned: 1
Jun 21 17:28:14 2d4eefd35cce (sd-parse-elf)[7457]: sym_dwfl_module_getelf for /usr/lib64/libc.so.6 returned NULL
Jun 21 17:28:14 2d4eefd35cce (sd-parse-elf)[7457]: parse_package_metadata for /usr/lib64/libc.so.6, with fallback method which consist in fetching using only the core dump file data returned: 0
Jun 21 17:28:14 2d4eefd35cce (sd-parse-elf)[7457]: sym_dwfl_module_getelf for /usr/lib64/libreadline.so.8.2 returned NULL
Jun 21 17:28:14 2d4eefd35cce (sd-parse-elf)[7457]: parse_package_metadata for /usr/lib64/libreadline.so.8.2, with fallback method which consist in fetching using only the core dump file data returned: 1
Jun 21 17:28:14 2d4eefd35cce (sd-parse-elf)[7457]: sym_dwfl_module_getelf for /usr/lib64/ld-linux-x86-64.so.2 returned NULL
Jun 21 17:28:14 2d4eefd35cce (sd-parse-elf)[7457]: parse_package_metadata for /usr/lib64/ld-linux-x86-64.so.2, with fallback method which consist in fetching using only the core dump file data returned: 0
Jun 21 17:28:14 2d4eefd35cce (sd-parse-elf)[7457]: sym_dwfl_module_getelf for linux-vdso.so.1 worked
Jun 21 17:28:14 2d4eefd35cce (sd-parse-elf)[7457]: parse_package_metadata for linux-vdso.so.1, with elf from sym_dwfl_module_getelf didn't return anything.
Jun 21 17:28:14 2d4eefd35cce systemd-coredump[7456]: [🡕] Process 59 (bash) of user 0 dumped core.

                                                     Module /usr/bin/sleep from rpm bash-5.2.15-3.fc38.x86_64
                                                     Module /usr/lib64/libtinfo.so.6.3 from rpm ncurses-6.3-5.20220501.fc37.x86_64
                                                     Module /usr/lib64/libreadline.so.8.2 from rpm readline-8.2-2.fc37.x86_64
                                                     Stack trace of thread 2:
                                                     #0  0x00007ff242961fc7 n/a (/usr/lib64/libc.so.6 + 0xd1fc7)
                                                     #1  0x000000003b975847 n/a (n/a + 0x0)
                                                     ELF object binary architecture: AMD x86-64
```

So there are some differences between both implementation. First we can ignore the inconstency of the first module (`Module /usr/bin/sleep from rpm bash-5.2.15-3.fc38.x86_64`), indeed the `bash` package doesn't provide the `sleep` command, but this inconsistency is explained above.

The current implementation reads that all other modules are provided by the `coreutils` package from fedora 38, while the core happened on a fedora 37. Obviously the `coreutils` package provides none of these modules. This is because the current implementation will stop at the very first PT_LOAD segment that it is able to extract a valid ELF header/program header array and from which it can extract a valid `.note.package` note. In a core the very first PT_LOAD is the mapping of the executable itself, in that case `sleep`, which is indeed provided by `coreutils`.

So this patch works, and actually fixes a real bug in systemd-coredump as reported modules are wrong. We also tested that actually this alone is enough to fix our performance issue that we tried to report and (wrongly) reproduce in https://sourceware.org/pipermail/elfutils-devel/2023q2/006225.html

Now I don't know if after this pull request we shall also try to invert the logic to find the elf headers of each module in a followup pull request. Indeed I think it's better to try to cope only with the core file itself, and extract notes from this file only, rather than calling libelf for each and every module, which will actually try to find the file via build id on the host, while technically the mount namespace of the host may be completely different from the mount namespace of the crashing process. That might also speed up systemd-coredump a bit by not doing any filesystem access to try to find modules.

bluca commented:
That's great, could you please put that info in the commit message? Then I'll just merge it right away, no need for new CI run for commit message updates

Romain-Geissler-1A commented:
Ok I have pushed a new commit message, less verbose than what I wrote here. I hope this is fine.

Romain-Geissler-1A reviewed (commented):
Seems like the message I wrote in review was never submitted to you, so here it is, "submitting" the review (with my answers). I am not used to reviews in the Github UI at all ! ;) (cf https://github.com/orgs/community/discussions/10369)
